### ISSUE TITLE
Expected Result.Success or Result.Failed applied to a single argument, got …

### ISSUE THREAD

**Issue by cpitclaudel** [FStarLang/FStar#1272]
This happens when processing `examples/pattern-matching/PatternMatching.fst` in the cpitclaudel_1272_repro branch:

```
ASSERTION FAILURE: Expected Result.Success or Result.Failed applied to a single argument, got (let () = FStar.Tactics.Types.tracepoint (FStar.Tactics.Types.incr_depth _) in
  FStar.Tactics.Effect._dm4f_TAC_bind_elab (PatternMatching.matching_problem *
      PatternMatching.abspat_continuation)
    Prims.unit
    _
    (reify (PatternMatching.matching_problem_of_abs (FStar.Tactics.Builtins.__embed (fun
                    a
                    b
                    h1
                    h2
                    uu___740030
                    ->
                    FStar.Tactics.Effect.bind (FStar.Tactics.Builtins.print (FStar.Reflection.Basic.inspect_bv 
                              h1))
                      (fun uu___109 ->
                          FStar.Tactics.Builtins.print (FStar.Reflection.Basic.inspect_bv h2))))))
    _
    (fun uu___740032 ->
        let #uu___737028, #uu___737029, #mp, #continuation = uu___740032 in
        FStar.Tactics.Effect._dm4f_TAC_bind_elab Prims.unit
          Prims.unit
          _
          (reify (FStar.Tactics.Builtins.print "AAA" ()))
          _
          (fun uu___740033 ->
              reify (FStar.Tactics.Builtins.exact (FStar.Tactics.Builtins.quote mp) ())))
    (FStar.Tactics.Types.decr_depth (FStar.Tactics.Types.incr_depth _)))
<:
FStar.Tactics.Result.__result Prims.unit
```

**Comment by mtzguido** [comment]
The normalizer halting there seems very weird. Can you recheck the example? I'm getting several undefined identifiers in it.

**Comment by cpitclaudel** [comment]
It does reproduce here.  The error happens on line 322:

```
let mp_example : matching_problem =
  synth_by_tactic
    (fun () ->
      let abs = quote (fun (a b: Type) (h1: hyp (a ==> b)) (h2: hyp (a)) (_: goal (squash b)) ->
                         let open FStar.Tactics in
                         print (inspect_bv h1);;
                         print (inspect_bv h2)) () in
      let mp, continuation = matching_problem_of_abs abs in
      print "AAA" ();
      // let qq = quote mp () in
      // print_term qq ();
      exact (quote mp) ())
```

**Comment by cpitclaudel** [comment]
Another repro in cpitclaudel_1272_repro_2.  I wonder why I'm running into this so much now :/

**Comment by cpitclaudel** [comment]
I think my install is very broken. This is failing:

```
module Minimal

open FStar.Tactics

let x () : Tot unit =
  assert_by_tactic True idtac
```

It only happens in interactive mode.  Let me reset my whole install.

**Comment by mtzguido** [comment]
I found a bug which could very well be the cause for this, so if you find a repro that would be great :)

**Comment by cpitclaudel** [comment]
Looking for it, but a full rebuild fixed it :/

**Comment by mtzguido** [comment]
Don't worry about it, I'm blowing up F* in new and exciting ways regardless :).

Some parts of your example started working, but I'm yet to do a full assessment.

**Comment by cpitclaudel** [comment]
Thanks!

**Comment by mtzguido** [comment]
I think this now works (in my branch). There were several fixes, and I'm not sure on which one gets this unstuck. The first repro now succeeds up to the first synth_by_tactic (and is ill-typed afterwards) and the second one works up to `example`, which fails since it seems (at a quick glance) that the tactic generates some ill-typed term.

Maybe you can check it out again and let me know if/how it explodes?

Also this used to abort without `--lax` while trying to encode some embedded binders, that works now.

**Comment by cpitclaudel** [comment]
Nice, thanks! Even better if you add a pointer to the issue in the commit message (it makes it easier when I read through commits) :)

**Comment by mtzguido** [comment]
Now in master. (but still unsure which commit fixed this). Please reopen if you find any related weirdness.

**Comment by cpitclaudel** [comment]
I just hit this again while running on master (with your fixed). Is it a different issue? Repro pushed to https://github.com/FStarLang/FStar/tree/cpitclaudel_1272_again

```
ASSERTION FAILURE: Expected Result.Success or Result.Failed applied to a single argument, got ((reify (PatternMatching.Mkmatching_solution (Prims.Cons<0> #(FStar.Pervasives.Native.tuple2<0, 0> Prims.string FStar.Reflection.Types.term) (FStar.Pervasives.Native.Mktuple2<0, 0> #Prims.string #FStar.Reflection.Types.term "b#772224" (FStar.Tactics.Builtins.__embed #Tm_unknown Prims.l_True)) (Prims.Cons<0> #(FStar.Pervasives.Native.tuple2<0, 0> Prims.string FStar.Reflection.Types.term) (FStar.Pervasives.Native.Mktuple2<0, 0> #Prims.string #FStar.Reflection.Types.term "a#772221" (FStar.Tactics.Builtins.__embed #Tm_unknown Prims.l_True)) (Prims.Nil<0> #.(FStar.Pervasives.Native.tuple2<0, 0> Prims.string FStar.Reflection.Types.term)))) (Prims.Cons<0> #(FStar.Pervasives.Native.tuple2<0, 0> Prims.string FStar.Reflection.Types.binder) (FStar.Pervasives.Native.Mktuple2<0, 0> #Prims.string #FStar.Reflection.Types.binder "h#772226" (Meta_alien "reflection.embed_binder")) (Prims.Nil<0> #.(FStar.Pervasives.Native.tuple2<0, 0> Prims.string FStar.Reflection.Types.binder)))) (Meta_alien "tactics.embed_proofstate")) <ascribed: [Prims.PURE] (FStar.Tactics.Result.__result<0> PatternMatching.matching_solution) )
```

**Comment by cpitclaudel** [comment]
Another repro in cpitclaudel_1272_ter

**Comment by cpitclaudel** [comment]
And another one in cpitclaudel_1272_quart
One thing that worries me is that we seem to be constructing gigantic terms: the error message this time is thousands of lines long

**Comment by mtzguido** [comment]
The second one looks different indeed, and very worrying! It seems it's trying to reify a record value, which cannot be well-typed. I'll take a look today, and at the others.

**Comment by cpitclaudel** [comment]
Ah, here's a much shorter one!

```
module Test

open FStar.Tactics

let unsquash #a : a -> squash a =
  fun _ -> ()

let broken (a: Type0) =
  assert_by_tactic a (apply (quote (unsquash #a));; let xx : a = admit() in exact (quote xx))

let yy : (Type0 -> unit) =
  synth_by_tactic (fun () -> exact (norm_term [] (quote broken ())) ())
```

```
ASSERTION FAILURE: Expected Result.Success or Result.Failed applied to a single argument, got reify ((((fun uu___672903 ->
            (let tt =
                (((fun uu___672906 ->
                        FStar.Tactics.Effect.TAC?.reflect (FStar.Tactics.Builtins.__norm_term Prims.Nil
                              (FStar.Tactics.Builtins.__embed Test.broken)))
                    <:
                    uu___672905:Prims.unit -> FStar.Tactics.Effect.Tac FStar.Reflection.Types.term)
                  <:
                  Prims.PURE
                  (uu___672907:Prims.unit -> FStar.Tactics.Effect.Tac FStar.Reflection.Types.term)) ()
                
                <:
                FStar.Tactics.Effect.TAC FStar.Reflection.Types.term
              in
              FStar.Tactics.Effect.TAC?.reflect (FStar.Tactics.Builtins.__exact tt))
            <:
            FStar.Tactics.Effect.TAC Prims.unit)
        <:
        uu___672902:Prims.unit -> FStar.Tactics.Effect.Tac Prims.unit)
      <:
      Prims.PURE (uu___672908:Prims.unit -> FStar.Tactics.Effect.Tac Prims.unit)) ())
  _
<:
FStar.Tactics.Result.__result Prims.unit
```

**Comment by mtzguido** [comment]
Thanks, that's great. That one actually looks like the Crash.fst you had on Slack, because the normalizer got stuck at trying to `reify` a function. The logic for reifying actions seems wrong at a few points, including the fact that DM4F actions get unfolded even without an enclosing reify.

Did you file that one in github too? If not I can go fetch it and open a new issue.

**Comment by cpitclaudel** [comment]
Keeping both here sounds good. I don't think I filed the slack one.

**Comment by mtzguido** [comment]
Pasting the thread.
--------------------------------------------------

cpitclaudel [Oct 1st at 12:08 PM] 
in #tactics
Also, what does "unexpected: term was not an alien embedding" mean?


12 replies
guido [8 days ago] 
This means we wanted to unembed a term/environment/proofstate but it wasn't what we expected. This shouldn't ever happen on well-typed tactics. Got the example?


cpitclaudel [8 days ago] 
Sure; not quite minimal yet, but:
```module Crash

open FStar.Tactics

type qn = list string
type var = string

type pattern =
| SPAny: pattern
| SPVar: var: var -> pattern
| SPQn: qn: qn -> pattern
| SPApp: hd: pattern -> arg: pattern -> pattern

noeq type match_exception =
| NameMismatch of qn * qn
| SimpleMismatch of pattern * term
| NonLinearMismatch of var * term * term
| UnsupportedTermInPattern of term

let string_of_match_exception = function _ -> ""

noeq type match_res a =
| Success of a
| Error of match_exception

let dm4f_match_res (a:Type) = unit -> M (match_res a)

let return_match_res a (x: a) : dm4f_match_res a =
  fun _ -> Success x

let bind_match_res (a b: Type)
                   (f: dm4f_match_res a)
                   (g: a -> dm4f_match_res b)
    : dm4f_match_res b =
  fun _ ->
    let r = f () in
    match r with
    | Success aa -> g aa ()
    | Error ex -> Error ex

let raise_match_res a (ex: match_exception) : dm4f_match_res a =
  fun _ -> Error ex

reifiable reflectable new_effect {
  EXN : (a:Type) -> Effect
  with repr     = dm4f_match_res
     ; bind     = bind_match_res
     ; return   = return_match_res
     ; raise (#a:Type) = raise_match_res a
}

let raise =
  EXN?.raise
effect Exn (a:Type) (pre:Type0) (post:EXN?.post a) =
  EXN a (fun (_:unit) (p:EXN?.post a) -> pre /\
          (forall (r: match_res a). (pre /\ post r) ==> p r))
effect Ex (a:Type) =
  EXN a (fun _ p -> forall x. p x)

let lift_exn_tac #a #b (f: a -> Ex b) (aa: a) : Tac b =
  match reify (f aa) () with
  | Success bb -> Tactics.return bb ()
  | Error ex -> Tactics.fail (string_of_match_exception ex) ()

type bindings = list (var * term)

let rec interp_pattern_tr
        (pat: pattern) (tm: term) (cur_bindings: bindings)
    : Ex bindings =
  match pat with
  | SPAny ->
    []
  | SPVar var ->
    (match List.Tot.assoc var cur_bindings with
     | Some tm' -> if term_eq tm tm' then cur_bindings
                  else raise (NonLinearMismatch (var, tm, tm'))
     | None -> ((var, tm) :: cur_bindings))
  | SPQn qn -> begin
    match inspect tm with
    | Tv_FVar fv ->
      if inspect_fv fv = qn then cur_bindings
      else raise (NameMismatch (qn, (inspect_fv fv)))
    | _ -> raise (SimpleMismatch (pat, tm)) end
  | SPApp p_hd p_arg -> begin
    match inspect tm with
    | Tv_App hd (arg, _) ->
      let with_hd = interp_pattern_tr p_hd hd cur_bindings in
      let with_arg = interp_pattern_tr p_arg arg with_hd in
      with_arg
    | _ -> raise (SimpleMismatch (pat, tm)) end

let interp_pattern (pat: pattern) (tm: term) : Ex bindings =
  List.Tot.rev (interp_pattern_tr pat tm [])

let rec pattern_of_term_ex tm : Ex pattern =
  match inspect tm with
  | Tv_Var bv -> SPVar (inspect_bv bv)
  | Tv_FVar fv -> SPQn (inspect_fv fv)
  | Tv_App f (x, _) ->
    SPApp (pattern_of_term_ex f) (pattern_of_term_ex x)
  | _ -> raise (UnsupportedTermInPattern tm)

let beta_reduce (tm: term) : Tac term =
  norm_term [] tm ()

let pattern_of_term tm : Tac pattern =
  let tm' = beta_reduce tm in
  lift_exn_tac pattern_of_term_ex tm'

let rec binders_and_body_of_abs tm : binders * term =
  match inspect tm with
  | Tv_Abs binder tm ->
    let binders, body = binders_and_body_of_abs tm in
    binder :: binders, body
  | _ -> [], tm

let pattern_of_abs tm : Tac pattern =
  let binders, body = binders_and_body_of_abs tm in
  pattern_of_term body

let match_term pat tm : Tac bindings =
  lift_exn_tac (interp_pattern pat) tm

let match_goal pat : Tac bindings =
  match_term pat (cur_goal ())

let test () =
  assert_by_tactic (1 == 1)
    (fun () ->
      let pat = quote (fun (x: int) -> x == x) () in
      let compiled = pattern_of_abs pat in
      let bindings = match_goal compiled in
      trefl ())
```
(edited)


cpitclaudel [8 days ago] 
You need to lax-check it until the test () part

cpitclaudel [8 days ago] 
Otherwise Z3 rejects it

guido [7 days ago] 
I could get z3 to accept this, and it still blows up. It has something to do with the definition of `raise`, because the normalizer gets stuck there

guido [7 days ago] 
uh-oh, actually it looks like something being reified twice...

guido [7 days ago] 
Lost all juice on the laptop, but seems that in the `lift_exn_tac` (name might not be exactly that) the inner part is already reified during normalization, and thus gets stuck

guido [7 days ago] 
I guess it might be the first use case for nested dm4f effects

guido [7 days ago] 
So the normalizer might need a kick

guido [7 days ago] 
If anyone wants to take a look, I first removed the new tracing things from the tactic bind to get a proper error

cpitclaudel [7 days ago] 
Not sure it's a great use case, though :) I think it'd be even better if exceptions were a subeffect of tactics.

cpitclaudel [7 days ago] 
(the way this is written, you can't use tactic debugging to step through the exception-raising function, for example)

**Comment by cpitclaudel** [comment]
I pushed another repro to cpitclaudel_1272_quint; it fails on the first use of `tfirst`.

The error message is this:

```
ASSERTION FAILURE: Expected Result.Success or Result.Failed applied to a single argument, got ((match (match (match (match (match (reify ((MonadicLift-Tm_abs{(binder:binder -> Tac (abspat_argspec)) : PURE -> TAC} (fun binder -> (((Monadic-Tm_delayed{TAC abspat_argspec} let  uu___#1399332  : (tuple2 abspat_binder_kind term) = (Monadic-Tm_delayed{TAC (tuple2 abspat_binder_kind term)} ((Monadic-Tm_delayed{TAC (tuple2 abspat_binder_kind term)} (match (match (match (match (__inspect (__type_of_binder binder@0:Tm_fvar: absvar)) with
	| (Tv_App hd#1399333 (Mktuple2 #.uu___#686770 #.uu___#686771 arg#1399334 _))  -> (bind #bindings #bindings (interp_pattern_aux (PQn (Cons #string "PatternMatching" (Cons #string "var" (Nil #.string)))) (Nil #.(tuple2 varname term)) hd@2:Tm_fvar: term) (fun with_hd -> ((bind #bindings #bindings (interp_pattern_aux (PVar "v") with_hd@0:Tm_fvar: bindings arg@2:Tm_fvar: term) (fun with_arg -> ((return #bindings with_arg@0:Tm_fvar: bindings) $$ (residual) Prims.Tot (match_res bindings)))) $$ (residual) Prims.Tot (match_res bindings))))
	|_  -> (raise #bindings (SimpleMismatch (Mktuple2 #pattern #term (PApp (PQn (Cons #string "PatternMatching" (Cons #string "var" (Nil #.string)))) (PVar "v")) (__type_of_binder binder@1:Tm_fvar: absvar))))) with
	| (Success #.uu___#680306 aa#1399339)  -> (return #(list (tuple2 varname term)) (rev #(tuple2 varname term) aa@0:Tm_bvar: a@3))
	|(Failure #.uu___#680445 ex#1399340)  -> (Failure #(list (tuple2 varname term)) ex@0:Tm_fvar: match_exception)) with
	| (Success #.uu___#697868 (Cons #.uu___#697865 (Mktuple2 #.uu___#697858 #.uu___#697859 _ var_typ#1399342) (Nil #.uu___#697861)))  -> (Success #(tuple2 abspat_binder_kind term) (Mktuple2 #abspat_binder_kind #term (ABKVar var_typ@0:Tm_fvar: term) var_typ@0:Tm_fvar: term))
	|(Failure #.uu___#698678 _)  -> ((match (interp_pattern (PApp (PQn hyp_qn) (PVar "v")) (__type_of_binder binder@1:Tm_fvar: absvar)) with
	| (Success #.uu___#698833 (Cons #.uu___#698830 (Mktuple2 #.uu___#698823 #.uu___#698824 _ hyp_typ#1399345) (Nil #.uu___#698826)))  -> (Success #(tuple2 abspat_binder_kind term) (Mktuple2 #abspat_binder_kind #term ABKHyp hyp_typ@0:Tm_fvar: term))
	|(Failure #.uu___#699774 _)  -> ((match (interp_pattern (PApp (PQn goal_qn) (PVar "v")) (__type_of_binder binder@2:Tm_fvar: absvar)) with
	| (Success #.uu___#699955 (Cons #.uu___#699952 (Mktuple2 #.uu___#699945 #.uu___#699946 _ goal_typ#1399348) (Nil #.uu___#699948)))  -> (Success #(tuple2 abspat_binder_kind term) (Mktuple2 #abspat_binder_kind #term ABKGoal goal_typ@0:Tm_fvar: term))
	|(Failure #.uu___#701048 _)  -> (Failure #(tuple2 abspat_binder_kind term) IncorrectTypeInAbsPatBinder)) <ascribed: [Prims.PURE] (match_res (tuple2 abspat_binder_kind term)) )) <ascribed: [Prims.PURE] (match_res (tuple2 abspat_binder_kind term)) )) with
	| (Success #.uu___#680819 bb#1399350)  -> (MonadicLift-Tm_bvar: bb@0{(tuple2 abspat_binder_kind term) : PURE -> TAC} bb@0:Tm_bvar: b@2)
	|(Failure #.uu___#680950 ex#1399351)  -> (Monadic-Tm_delayed{TAC (tuple2 abspat_binder_kind term)} (fail #(tuple2 abspat_binder_kind term) (string_of_match_exception ex@0:Tm_fvar: match_exception) ())))) <ascribed: [FStar.Tactics.Effect.Tac] (tuple2 abspat_binder_kind term) ))
in
(MonadicLift-Tm_delayed{abspat_argspec : PURE -> TAC} ((match uu___@0:Tm_delayed with
	| (Mktuple2 #.uu___#705159 #.uu___#705160 binder_kind#1399352 _)  -> (Mkabspat_argspec binder@3:Tm_fvar: absvar binder_kind@1:Tm_fvar: abspat_binder_kind)) <ascribed: [Prims.PURE] abspat_argspec ))) <ascribed: [FStar.Tactics.Effect.Tac] abspat_argspec ) $$ (residual) FStar.Tactics.Effect.Tac abspat_argspec))) (Meta_alien "reflection.embed_binder")) (Meta_alien "tactics.embed_proofstate")) with
	| (Success #.uu___#566892 (Mktuple2 #.uu___#566888 #.uu___#566889 a#1399354 q#1399355))  -> ((match (tracepoint q@0:Tm_fvar: proofstate) with
	| ()  -> (_dm4f_TAC_bind_elab (list abspat_argspec) (list abspat_argspec) Tm_unknown (reify (tacmap #binder #abspat_argspec (MonadicLift-Tm_delayed{(binder:binder -> Tac (abspat_argspec)) : PURE -> TAC} (fun binder -> (((Monadic-Tm_delayed{TAC abspat_argspec} let  uu___#1399332  : (tuple2 abspat_binder_kind term) = (Monadic-Tm_delayed{TAC (tuple2 abspat_binder_kind term)} ((Monadic-Tm_delayed{TAC (tuple2 abspat_binder_kind term)} (match (match (match (match (__inspect (__type_of_binder binder@0:Tm_fvar: absvar)) with
	| (Tv_App hd#1399333 (Mktuple2 #.uu___#686770 #.uu___#686771 arg#1399334 _))  -> (bind #bindings #bindings (interp_pattern_aux (PQn (Cons #string "PatternMatching" (Cons #string "var" (Nil #.string)))) (Nil #.(tuple2 varname term)) hd@2:Tm_fvar: term) (fun with_hd -> ((bind #bindings #bindings (interp_pattern_aux (PVar "v") with_hd@0:Tm_fvar: bindings arg@2:Tm_fvar: term) (fun with_arg -> ((return #bindings with_arg@0:Tm_fvar: bindings) $$ (residual) Prims.Tot (match_res bindings)))) $$ (residual) Prims.Tot (match_res bindings))))
	|_  -> (raise #bindings (SimpleMismatch (Mktuple2 #pattern #term (PApp (PQn (Cons #string "PatternMatching" (Cons #string "var" (Nil #.string)))) (PVar "v")) (__type_of_binder binder@1:Tm_fvar: absvar))))) with
	| (Success #.uu___#680306 aa#1399339)  -> (return #(list (tuple2 varname term)) (rev #(tuple2 varname term) aa@0:Tm_bvar: a@3))
	|(Failure #.uu___#680445 ex#1399340)  -> (Failure #(list (tuple2 varname term)) ex@0:Tm_fvar: match_exception)) with
	| (Success #.uu___#697868 (Cons #.uu___#697865 (Mktuple2 #.uu___#697858 #.uu___#697859 _ var_typ#1399342) (Nil #.uu___#697861)))  -> (Success #(tuple2 abspat_binder_kind term) (Mktuple2 #abspat_binder_kind #term (ABKVar var_typ@0:Tm_fvar: term) var_typ@0:Tm_fvar: term))
	|(Failure #.uu___#698678 _)  -> ((match (interp_pattern (PApp (PQn hyp_qn) (PVar "v")) (__type_of_binder binder@1:Tm_fvar: absvar)) with
	| (Success #.uu___#698833 (Cons #.uu___#698830 (Mktuple2 #.uu___#698823 #.uu___#698824 _ hyp_typ#1399345) (Nil #.uu___#698826)))  -> (Success #(tuple2 abspat_binder_kind term) (Mktuple2 #abspat_binder_kind #term ABKHyp hyp_typ@0:Tm_fvar: term))
	|(Failure #.uu___#699774 _)  -> ((match (interp_pattern (PApp (PQn goal_qn) (PVar "v")) (__type_of_binder binder@2:Tm_fvar: absvar)) with
	| (Success #.uu___#699955 (Cons #.uu___#699952 (Mktuple2 #.uu___#699945 #.uu___#699946 _ goal_typ#1399348) (Nil #.uu___#699948)))  -> (Success #(tuple2 abspat_binder_kind term) (Mktuple2 #abspat_binder_kind #term ABKGoal goal_typ@0:Tm_fvar: term))
	|(Failure #.uu___#701048 _)  -> (Failure #(tuple2 abspat_binder_kind term) IncorrectTypeInAbsPatBinder)) <ascribed: [Prims.PURE] (match_res (tuple2 abspat_binder_kind term)) )) <ascribed: [Prims.PURE] (match_res (tuple2 abspat_binder_kind term)) )) with
	| (Success #.uu___#680819 bb#1399350)  -> (MonadicLift-Tm_bvar: bb@0{(tuple2 abspat_binder_kind term) : PURE -> TAC} bb@0:Tm_bvar: b@2)
	|(Failure #.uu___#680950 ex#1399351)  -> (Monadic-Tm_delayed{TAC (tuple2 abspat_binder_kind term)} (fail #(tuple2 abspat_binder_kind term) (string_of_match_exception ex@0:Tm_fvar: match_exception) ())))) <ascribed: [FStar.Tactics.Effect.Tac] (tuple2 abspat_binder_kind term) ))
in
(MonadicLift-Tm_delayed{abspat_argspec : PURE -> TAC} ((match uu___@0:Tm_delayed with
	| (Mktuple2 #.uu___#705159 #.uu___#705160 binder_kind#1399352 _)  -> (Mkabspat_argspec binder@3:Tm_fvar: absvar binder_kind@1:Tm_fvar: abspat_binder_kind)) <ascribed: [Prims.PURE] abspat_argspec ))) <ascribed: [FStar.Tactics.Effect.Tac] abspat_argspec ) $$ (residual) FStar.Tactics.Effect.Tac abspat_argspec))) (Cons #binder (Meta_alien "reflection.embed_binder") (Nil #.binder)))) Tm_unknown (fun uu___ -> ((_dm4f_TAC_return_elab (list abspat_argspec) (Cons #abspat_argspec a@2:Tm_bvar: a@7 uu___@0:Tm_delayed)) $$ (residual) FStar.Tactics.Effect.TAC (list (MonadicLift-Tm_delayed{(binder:binder -> Tac (abspat_argspec)) : PURE -> TAC} (fun binder -> (((Monadic-Tm_delayed{TAC abspat_argspec} let  uu___#1399332  : (tuple2 abspat_binder_kind term) = (Monadic-Tm_delayed{TAC (tuple2 abspat_binder_kind term)} ((Monadic-Tm_delayed{TAC (tuple2 abspat_binder_kind term)} (match (match (match (match (__inspect (__type_of_binder binder@0:Tm_fvar: absvar)) with
	| (Tv_App hd#1399333 (Mktuple2 #.uu___#686770 #.uu___#686771 arg#1399334 _))  -> (bind #bindings #bindings (interp_pattern_aux (PQn (Cons #string "PatternMatching" (Cons #string "var" (Nil #.string)))) (Nil #.(tuple2 varname term)) hd@2:Tm_fvar: term) (fun with_hd -> ((bind #bindings #bindings (interp_pattern_aux (PVar "v") with_hd@0:Tm_fvar: bindings arg@2:Tm_fvar: term) (fun with_arg -> ((return #bindings with_arg@0:Tm_fvar: bindings) $$ (residual) Prims.Tot (match_res bindings)))) $$ (residual) Prims.Tot (match_res bindings))))
	|_  -> (raise #bindings (SimpleMismatch (Mktuple2 #pattern #term (PApp (PQn (Cons #string "PatternMatching" (Cons #string "var" (Nil #.string)))) (PVar "v")) (__type_of_binder binder@1:Tm_fvar: absvar))))) with
	| (Success #.uu___#680306 aa#1399339)  -> (return #(list (tuple2 varname term)) (rev #(tuple2 varname term) aa@0:Tm_bvar: a@3))
	|(Failure #.uu___#680445 ex#1399340)  -> (Failure #(list (tuple2 varname term)) ex@0:Tm_fvar: match_exception)) with
	| (Success #.uu___#697868 (Cons #.uu___#697865 (Mktuple2 #.uu___#697858 #.uu___#697859 _ var_typ#1399342) (Nil #.uu___#697861)))  -> (Success #(tuple2 abspat_binder_kind term) (Mktuple2 #abspat_binder_kind #term (ABKVar var_typ@0:Tm_fvar: term) var_typ@0:Tm_fvar: term))
	|(Failure #.uu___#698678 _)  -> ((match (interp_pattern (PApp (PQn hyp_qn) (PVar "v")) (__type_of_binder binder@1:Tm_fvar: absvar)) with
	| (Success #.uu___#698833 (Cons #.uu___#698830 (Mktuple2 #.uu___#698823 #.uu___#698824 _ hyp_typ#1399345) (Nil #.uu___#698826)))  -> (Success #(tuple2 abspat_binder_kind term) (Mktuple2 #abspat_binder_kind #term ABKHyp hyp_typ@0:Tm_fvar: term))
	|(Failure #.uu___#699774 _)  -> ((match (interp_pattern (PApp (PQn goal_qn) (PVar "v")) (__type_of_binder binder@2:Tm_fvar: absvar)) with
	| (Success #.uu___#699955 (Cons #.uu___#699952 (Mktuple2 #.uu___#699945 #.uu___#699946 _ goal_typ#1399348) (Nil #.uu___#699948)))  -> (Success #(tuple2 abspat_binder_kind term) (Mktuple2 #abspat_binder_kind #term ABKGoal goal_typ@0:Tm_fvar: term))
	|(Failure #.uu___#701048 _)  -> (Failure #(tuple2 abspat_binder_kind term) IncorrectTypeInAbsPatBinder)) <ascribed: [Prims.PURE] (match_res (tuple2 abspat_binder_kind term)) )) <ascribed: [Prims.PURE] (match_res (tuple2 abspat_binder_kind term)) )) with
	| (Success #.uu___#680819 bb#1399350)  -> (MonadicLift-Tm_bvar: bb@0{(tuple2 abspat_binder_kind term) : PURE -> TAC} bb@0:Tm_bvar: b@2)
	|(Failure #.uu___#680950 ex#1399351)  -> (Monadic-Tm_delayed{TAC (tuple2 abspat_binder_kind term)} (fail #(tuple2 abspat_binder_kind term) (string_of_match_exception ex@0:Tm_fvar: match_exception) ())))) <ascribed: [FStar.Tactics.Effect.Tac] (tuple2 abspat_binder_kind term) ))
in
(MonadicLift-Tm_delayed{abspat_argspec : PURE -> TAC} ((match uu___@0:Tm_delayed with
	| (Mktuple2 #.uu___#705159 #.uu___#705160 binder_kind#1399352 _)  -> (Mkabspat_argspec binder@3:Tm_fvar: absvar binder_kind@1:Tm_fvar: abspat_binder_kind)) <ascribed: [Prims.PURE] abspat_argspec ))) <ascribed: [FStar.Tactics.Effect.Tac] abspat_argspec ) $$ (residual) FStar.Tactics.Effect.Tac abspat_argspec)))))) (decr_depth q@0:Tm_fvar: proofstate))) <ascribed: [Prims.PURE] (__result (list abspat_argspec)) )
	|(Failed #.uu___#567520 (Mktuple2 #.uu___#567516 #.uu___#567517 msg#1399363 q#1399364))  -> (Failed #(list abspat_argspec) (Mktuple2 #string #proofstate msg@1:Tm_fvar: string q@0:Tm_fvar: proofstate))) with
	| (Success #.uu___#566892 (Mktuple2 #.uu___#566888 #.uu___#566889 a#1399365 q#1399366))  -> ((match (tracepoint q@0:Tm_fvar: proofstate) with
	| ()  -> (_dm4f_TAC_return_elab (tuple2 (list abspat_argspec) term) (Mktuple2 #(list abspat_argspec) #term a@1:Tm_bvar: a@7 (__embed #((fun a x uu___ -> (a@2:Tm_type $$ (residual) Prims.Tot Type)) ((fun a abspat -> (a@1:Tm_type $$ (residual) Prims.Tot Type)) (a:(var Type) -> h:(hyp (squash a@0:Tm_type)) -> Tac (unit)) (fun a h -> (((Monadic-Tm_delayed{TAC unit} ((clear h@0:Tm_fvar: binder) ())) <ascribed: [FStar.Tactics.Effect.TAC] unit ) $$ (residual) FStar.Tactics.Effect.Tac unit))) (fun a h -> (((Monadic-Tm_delayed{TAC unit} ((clear h@0:Tm_fvar: binder) ())) <ascribed: [FStar.Tactics.Effect.TAC] unit ) $$ (residual) FStar.Tactics.Effect.Tac unit)) ()) (fun a h -> (((Monadic-Tm_delayed{TAC unit} ((clear h@0:Tm_fvar: binder) ())) <ascribed: [FStar.Tactics.Effect.TAC] unit ) $$ (residual) FStar.Tactics.Effect.Tac unit)))) (decr_depth q@0:Tm_fvar: proofstate))) <ascribed: [Prims.PURE] (__result (tuple2 (list abspat_argspec) term)) )
	|(Failed #.uu___#567520 (Mktuple2 #.uu___#567516 #.uu___#567517 msg#1399367 q#1399368))  -> (Failed #(tuple2 (list abspat_argspec) term) (Mktuple2 #string #proofstate msg@1:Tm_fvar: string q@0:Tm_fvar: proofstate))) with
	| (Success #.uu___#566892 (Mktuple2 #.uu___#566888 #.uu___#566889 a#1399369 q#1399370))  -> ((match (tracepoint q@0:Tm_fvar: proofstate) with
	| ()  -> (_dm4f_TAC_return_elab (tuple2 matching_problem abspat_continuation) (Mktuple2 #matching_problem #abspat_continuation (Mkmatching_problem (rev #varname (__proj__Mkmatching_problem__item__mp_vars (Mkmatching_problem (Cons #string (MonadicLift-Tm_constant{string : PURE -> TAC} "a#1399291") (Nil #.string)) (Cons #(tuple2 string pattern) (Mktuple2 #string #pattern (MonadicLift-Tm_constant{string : PURE -> TAC} "h#1399293") (PApp (PQn (Cons #string "Prims" (Cons #string "squash" (Nil #string)))) (PVar "a#1399291"))) (Nil #.(tuple2 string pattern))) (None #.pattern)))) (rev #(tuple2 varname pattern) (__proj__Mkmatching_problem__item__mp_hyps (Mkmatching_problem (Cons #string (MonadicLift-Tm_constant{string : PURE -> TAC} "a#1399291") (Nil #.string)) (Cons #(tuple2 string pattern) (Mktuple2 #string #pattern (MonadicLift-Tm_constant{string : PURE -> TAC} "h#1399293") (PApp (PQn (Cons #string "Prims" (Cons #string "squash" (Nil #string)))) (PVar "a#1399291"))) (Nil #.(tuple2 string pattern))) (None #.pattern)))) (__proj__Mkmatching_problem__item__mp_goal (Mkmatching_problem (Cons #string (MonadicLift-Tm_constant{string : PURE -> TAC} "a#1399291") (Nil #.string)) (Cons #(tuple2 string pattern) (Mktuple2 #string #pattern (MonadicLift-Tm_constant{string : PURE -> TAC} "h#1399293") (PApp (PQn (Cons #string "Prims" (Cons #string "squash" (Nil #string)))) (PVar "a#1399291"))) (Nil #.(tuple2 string pattern))) (None #.pattern)))) a@1:Tm_bvar: a@7) (decr_depth q@0:Tm_fvar: proofstate))) <ascribed: [Prims.PURE] (__result (tuple2 matching_problem abspat_continuation)) )
	|(Failed #.uu___#567520 (Mktuple2 #.uu___#567516 #.uu___#567517 msg#1399371 q#1399372))  -> (Failed #(tuple2 matching_problem abspat_continuation) (Mktuple2 #string #proofstate msg@1:Tm_fvar: string q@0:Tm_fvar: proofstate))) with
	| (Success #.uu___#566892 (Mktuple2 #.uu___#566888 #.uu___#566889 a#1399373 q#1399374))  -> ((match (tracepoint q@0:Tm_fvar: proofstate) with
	| ()  -> ((match a@1:Tm_bvar: a@7 with
	| (Mktuple2 #.uu___#726364 #.uu___#726365 problem#1399375 continuation#1399376)  -> (_dm4f_TAC_bind_elab (uu___:matching_solution -> Tac (matching_solution)) matching_solution Tm_unknown (reify (tpair #abspat_continuation #matching_solution continuation@0:Tm_fvar: abspat_continuation)) Tm_unknown (fun uu___ -> ((reify (solve_mp #matching_solution problem@2:Tm_fvar: matching_problem (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Cons #binder (Meta_alien "reflection.embed_binder") (Nil #binder)))))))))))))))))))))))))))))))))))))))))))))))) (__embed #Tm_unknown (squash (l_imp (l_and a b) (l_imp (eq2 #((fun a b c -> (Type $$ (residual) Prims.Tot Type)) a b c) c b) c)))) uu___@0:Tm_delayed)) $$ (residual) FStar.Tactics.Effect.TAC matching_solution)))) (decr_depth q@0:Tm_fvar: proofstate))) <ascribed: [Prims.PURE] (__result (tuple2 (tuple2 (list abspat_argspec) term) matching_solution)) )
	|(Failed #.uu___#567520 (Mktuple2 #.uu___#567516 #.uu___#567517 msg#1399380 q#1399381))  -> (Failed #(tuple2 (tuple2 (list abspat_argspec) term) matching_solution) (Mktuple2 #string #proofstate msg@1:Tm_fvar: string q@0:Tm_fvar: proofstate))) with
	| (Success #.uu___#566892 (Mktuple2 #.uu___#566888 #.uu___#566889 a#1399382 q#1399383))  -> ((match (tracepoint q@0:Tm_fvar: proofstate) with
	| ()  -> ((match a@1:Tm_delayed with
	| (Mktuple2 #.uu___#729568 #.uu___#729569 continuation#1399384 solution#1399385)  -> (_dm4f_TAC_bind_elab (uu___:matching_solution -> Tac (unit)) unit Tm_unknown (reify (interp_abspat_continuation unit continuation@1:Tm_fvar: abspat_continuation)) Tm_unknown (fun uu___ -> ((reify (uu___@0:Tm_delayed solution@1:Tm_fvar: matching_solution)) $$ (residual) FStar.Tactics.Effect.TAC unit)))) (decr_depth q@0:Tm_fvar: proofstate))) <ascribed: [Prims.PURE] (__result unit) )
	|(Failed #.uu___#567520 (Mktuple2 #.uu___#567516 #.uu___#567517 msg#1399389 q#1399390))  -> (Failed #unit (Mktuple2 #string #proofstate msg@1:Tm_fvar: string q@0:Tm_fvar: proofstate))) <ascribed: [Prims.PURE] (__result unit) )
```

I had to use `--ugly` because otherwise F* crashed before that, and said "not an imp"

**Comment by mtzguido** [comment]
This is hairy, and the part of the compiler I probably know the least about, but here's a workaround for your smaller example, in case it helps unblock you.
```
let yy : (Type0 -> unit) =
  synth_by_tactic (fun () -> let t = quote broken () in exact (norm_term [] t) ())
```
Seems the `let` is better behaved than that effectful application.

**Comment by mtzguido** [comment]
Alas, I present my wonderful fix to this issue:
```
-    let tactic, _, g = TcTerm.tc_reified_tactic env tactic in
+    let _, _, g = TcTerm.tc_reified_tactic env tactic in
```
c.f. #1307, typechecking twice messes up monadic annotations.

This fixes your minimized example, but I'm not absolutely sure it fixes all the others (most of them are now stale...), could you give it a try again and report back?

**Comment by mtzguido** [comment]
Optimistically closing..
